### PR TITLE
Fix repository version listing after version deletion

### PR DESCRIPTION
**Pull request by gerrod3**
fixes: #9481

Please be sure you have read our documentation on creating PRs:
https://docs.pulpproject.org/contributing/pull-request-walkthrough.html

### CODE REVIEW

File: pulpcore/app/models/repository.py [hunk 1]
```diff
@@ -909,6 +909,9 @@ def _squash(self, repo_relations, next_version):
         repo_relations.filter(version_added=self).update(version_added=next_version)
         repo_relations.filter(version_removed=self).update(version_removed=next_version)
 
+        # Update next version's counts as they have been modified
+        self._compute_counts(next_version)
```
Comment (mdellweg) [inline pulpcore/app/models/repository.py:913]: would `next_version._compute_counts()` have worked too?

**Comment by daviddavis** [comment]
I think this might also fix this bug I reported some months back https://pulp.plan.io/issues/8798. Or maybe they're the same bug.

File: pulpcore/app/models/repository.py (hunk 1 above)
Comment (gerrod3) [reply pulpcore/app/models/repository.py]: :facepalm:, yes.

**Comment by gerrod3** [comment]
> I think this might also fix this bug I reported some months back https://pulp.plan.io/issues/8798. Or maybe they're the same bug.

I think you are right. I'll add tests later to be sure this has been fixed.

File: pulpcore/tests/functional/api/using_plugin/test_repo_versions.py [hunk 2]
```diff
@@ -357,12 +357,16 @@ def test_delete_last_version(self):
 
     def test_delete_middle_version(self):
         """Delete a middle version."""
-        index = randint(1, len(self.repo_version_hrefs) - 2)
+        index = randint(1, len(self.repo_version_hrefs) - 3)
```
Comment (mdellweg) [inline pulpcore/tests/functional/api/using_plugin/test_repo_versions.py:360]: I know, not introduced here, but using randomness in tests is really a two sided sword. You may eventually cover more cases, but you set the system up for a blinker, so it does not make a good regression. I'd rather pick a number.

**Review by mdellweg — APPROVED** [review]
LGTM

Just wondering that a single line is supposed to fix two different issues. Should they be marked duplicated instead?

**Comment by ipanova** [comment]
> LGTM
> 
> Just wondering that a single line is supposed to fix two different issues. Should they be marked duplicated instead?

not having a strong opinion on this, but probably closing one as dup would be my preference.

**Review by ipanova — APPROVED** [review]

File: pulpcore/tests/functional/api/using_plugin/test_repo_versions.py (hunk 2 above)
Comment (gerrod3) [reply pulpcore/tests/functional/api/using_plugin/test_repo_versions.py:360]: True, but currently the test sets up each repo version as adding only one content unit, so I'm just making sure I pick a version that has another one above it, however adding on more to this test is difficult.

**Comment by gerrod3** [comment]
> > LGTM
> > Just wondering that a single line is supposed to fix two different issues. Should they be marked duplicated instead?
> 
> not having a strong opinion on this, but probably closing one as dup would be my preference.

Which one should I close? 9481 will occur when a content type is no longer present after repo version deletion because the counts weren't updated. 8798 occurs every time you delete a previous repo version. 9481 was reported by a user, 8798 was noticed by David.

**Comment by mdellweg** [comment]
> > > LGTM
> > > Just wondering that a single line is supposed to fix two different issues. Should they be marked duplicated instead?
> > 
> > 
> > not having a strong opinion on this, but probably closing one as dup would be my preference.
> 
> Which one should I close? 9481 will occur when a content type is no longer present after repo version deletion because the counts weren't updated. 8798 occurs every time you delete a previous repo version. 9481 was reported by a user, 8798 was noticed by David.

So they are not really duplicate errors. Then we should "Fix" them both. Does the test cover both cases? (I don't think so.) Would it be feasible to make such a test with the file plugin?

BTW: What happens if you add a content in one version, remove it the next and then delete the former version. Would this result in a version without changes, or would it burst into flames?

**Comment by gerrod3** [comment]
> So they are not really duplicate errors. Then we should "Fix" them both. Does the test cover both cases? (I don't think so.) Would it be feasible to make such a test with the file plugin?

No, the issue reported by the user can only occur in plugins with more than one content type. When you list the repository versions it gives their summary report which includes the content type, but this content type is grabbed dynamically from the database each time since the summary only stores the pulp_type. So if a content type is removed through repo version deletion with a corresponding orphan cleanup, then the server will 500 error when trying to list the repo versions.

> BTW: What happens if you add a content in one version, remove it the next and then delete the former version. Would this result in a version without changes, or would it burst into flames?

You get a new version without any changes. Should I add a check to see if the updated next version is now the same as the previous version and possibly delete this duplicate repo version if so?

**Comment by mdellweg** [comment]
> > BTW: What happens if you add a content in one version, remove it the next and then delete the former version. Would this result in a version without changes, or would it burst into flames?
> 
> You get a new version without any changes. Should I add a check to see if the updated next version is now the same as the previous version and possibly delete this duplicate repo version if so?

An empty changeset is probably the best we can do, because deleting the second version also would definitely break stuff. It may be referenced by a pulication/distribution, or even outside Pulp. It must not disappear without being explicitely deleted.

Having said that, i think this PR is ready to go.

**Pull request merged**